ouroboros-network/src/Ouroboros/Network/Protocol/Trans/Hello/Codec.hs

```diff
@@ -0,0 +1,191 @@
+import           Text.Printf
+
+
+byteLimitsHello
+    :: forall ps (stIdle :: ps) bytes .
+       ProtocolSizeLimits ps bytes
+    -> ProtocolSizeLimits (Hello ps stIdle) bytes
+byteLimitsHello ProtocolSizeLimits { sizeLimitForState, dataSize }
+    = ProtocolSizeLimits {
+          sizeLimitForState = sizeLimitForStateHello,
+          dataSize
+        }
+  where 
+    sizeLimitForStateHello :: forall (pr :: PeerRole) (st :: Hello ps stIdle).
+                              PeerHasAgency pr st -> Word
+    sizeLimitForStateHello (ClientAgency TokHello) =
+      smallByteLimit
+    sizeLimitForStateHello (ClientAgency (TokClientTalk tok)) =
+      sizeLimitForState (ClientAgency tok)
+    sizeLimitForStateHello (ServerAgency (TokServerTalk tok)) =
+      sizeLimitForState (ServerAgency tok)
+
+
+timeLimitsHello
+    :: forall ps (stIdle :: ps).
+       ProtocolTimeLimits ps
+    -> ProtocolTimeLimits (Hello ps stIdle)
+timeLimitsHello ProtocolTimeLimits { timeLimitForState } =
+    ProtocolTimeLimits {
+        timeLimitForState = timeLimitForStateHello
+      }
+  where
+    timeLimitForStateHello :: forall (pr :: PeerRole) (st :: Hello ps stIdle).
+                               PeerHasAgency pr st -> Maybe DiffTime
+    timeLimitForStateHello (ClientAgency TokHello) = waitForever
+    timeLimitForStateHello (ClientAgency (TokClientTalk tok)) =
+      timeLimitForState (ClientAgency tok)
+    timeLimitForStateHello (ServerAgency (TokServerTalk tok)) =
+      timeLimitForState (ServerAgency tok)
+
+
+-- | 'codecHello' provides a flat encoding of the original codec and 'MsgHello'.
+-- 'MsgTalk' is invisible.  It assumes that the top level encoding of @ps@
+-- protocol is a list of at least one element (the tag of a message).
+--
+codecHello
+    :: forall ps (stIdle :: ps) m.
+       ( MonadST m
+       , ShowProxy ps
+       , ShowProxy stIdle
+       , forall (st :: ps). Show (ClientHasAgency st)
+       , forall (st :: ps). Show (ServerHasAgency st)
+       )
+    => Word
+    -- ^ tag for 'MsgHello'
+    -> (forall (pr :: PeerRole) (st :: ps) (st' :: ps).
+               PeerHasAgency pr st
+            -> Message ps st st' -> CBOR.Encoding)
+    -> (forall (pr :: PeerRole) (st :: ps) s.
+               PeerHasAgency pr st
+            -> Int
+            -> Word
+            -> CBOR.Decoder s (SomeMessage st))
+    -- ^ continuation which receives agency, tag, and length. Last two are
+    -- decoded values.
+    -> Codec (Hello ps stIdle) CBOR.DeserialiseFailure m LBS.ByteString
+codecHello helloTag encode decode = mkCodecCborLazyBS encodeHello decodeHello
+  where
+    encodeHello :: forall (pr  :: PeerRole)
+                     (st  :: Hello ps stIdle)
+                     (st' :: Hello ps stIdle).
+                  PeerHasAgency pr st
+               -> Message (Hello ps stIdle) st st'
+               -> CBOR.Encoding
+    encodeHello (ClientAgency TokHello) MsgHello =
+      encodeListLen 1 <> encodeWord helloTag
```

Comment:
Or we could generalise over the whole hello `Encoding`, not just a `Word`. Doesn't matter at the moment.

Reply:
Unfortunately not, it would need to go through `Codec.Cbor.Term.Term` or something like that: the decoder is in CPS style.